src/IW/Sync/Search.hs

```diff
@@ -1,97 +1,195 @@
-    :: ( MonadIO m
+-- | Search all open issues with Haskell language and the labels passed in to the function.
+searchAllHaskellIssuesByLabels
+    :: forall env m.
+       ( MonadIO m
+       , MonadUnliftIO m
        , WithError m
        , WithLog env m
        )
-    => m [GitHub.Issue]
-fetchAllHaskellIssues = fetchHaskellIssuesByLabels []
+    => [Label] -- ^ Issue labels
+    -> Integer -- ^ Search interval
+    -> m [Issue]
+searchAllHaskellIssuesByLabels labels interval = do
+     today <- liftIO getToday
+     githubIssues <- githubSearchAll ["search", "issues"] queryString today interval []
+     pure $ catMaybes $ fromGitHubIssue <$> githubIssues
+  where
+    queryString :: Text
+    queryString = "language:haskell " <> labelsToSearchQuery labels
 
--- | Lift a github search action to work within our monad stack.
-liftGitHubSearchToApp
+    -- | Construct a github search query from a list of labels.
+    labelsToSearchQuery :: [Label] -> Text
+    labelsToSearchQuery = foldMap (\Label{..} -> "label:\"" <> unLabel <> "\" ")
+
+githubSearchAll
     :: forall a env m.
        ( MonadIO m
+       , MonadUnliftIO m
        , WithError m
        , WithLog env m
+       , FromJSON a
        , Typeable a
        )
-    => IO (Either GitHub.Error (SearchResult a))
+    => Paths   -- ^ URL paths
+    -> Text    -- ^ Query properties
+    -> Day     -- ^ The function starts at this day and goes back in time
+    -> Integer -- ^ Interval
+    -> [a]     -- ^ Accumulated results used in recursive calls to this function
     -> m [a]
-liftGitHubSearchToApp githubSearch = liftIO githubSearch >>= \case
-    Left err -> throwError $ githubErrToAppErr err
-    Right (SearchResult count vec) -> do
-        log Info $ "Fetching total of " <> show count <> " " <> typeName @a <> "s..."
-        pure $ V.toList vec
+githubSearchAll paths properties recent interval acc = do
+    let startDay = negate interval `addDays` recent
+    (SearchResult count vec) <- liftGithubSearchToApp $ githubSearch paths properties startDay recent 1
+        -- | If count is 0, then all results for the given properties have been obtained.
+    if | count == 0   -> do
+            log I "All results obtained."
+            pure acc
+        -- | If count is less than 1000, then the interval is good and we can get
+        -- the rest of the results on pages 2 to 10.
+       | count < 1000 -> do
+            remResults <- foldMapM (githubSearch' paths properties startDay recent) [2..10]
+            githubSearchAll paths properties (pred startDay) interval (acc <> V.toList vec <> remResults)
+        -- | Otherwise, call the function with the same arguments but a smaller interval.
+       | otherwise    -> githubSearchAll paths properties recent (pred interval) acc
 
--- | Search all repositories built with the Haskell language.
-searchHaskellRepos :: IO (Either GitHub.Error (SearchResult GitHub.Repo))
-searchHaskellRepos = searchRepos "language:haskell"
+githubSearch'
```

Comment:
I personally don't like `'` in names just to have different functions. Can we rename this function or `githubSearch` to something else to have clear difference between functions?

Reply:
I agree completely. Couldn’t think of a name so that’s just a placeholder until I come up with something better. I hate getting stuck st naming functions😂